### ISSUE TITLE
Verify pact in provider side with gradle

### ISSUE THREAD

**Issue by johnelf** [DiUS/pact-jvm#86]
Hi guys, I want to verify my pact file at provider side, and I use gradle as management tool, so here's my code snippet.
java version: 8
gradle version: 1.10
```gradle
buildscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'au.com.dius:pact-jvm-provider-gradle_2.10:2.1.9'
    }
}
apply plugin: 'au.com.dius.pact'

 pact {
        serviceProviders {
            listingServiceProvider { providerInfo ->
                host = 'localhost'
                port = 8080
                hasPactWith('saved properties dashboard') { consumerInfo ->
                    consumerInfo.pactFile = file('src/pactTest/resources/consumer-provider.json')
                    stateChange = url('http://localhost:8080')
                }
            }
        }
    }
```
but I got `State Change Request Failed - Not Found` error for all scenarios when checking the pact.
the pact file contains several interactions between consumer and provider, including POST and GET http method.

if I just remove stateChange, the pact seems not get run. :( any suggestions?

**Comment by johnelf** [comment]
BTW, if I remove stateChange, just like README.md
I got
```
State Change Request Failed - Target host must not be null, or set in parameters.
```

**Comment by uglyog** [comment]
The pact gradle plugin expects a stateChange URL to send the the provider states that are defined in the pact file. So when it finds an interaction with a provider state, it will POST that state to the URL you gave: `http://localhost:8080`. This POST request is failing, and the HTTP client library is returning `not found` as the error (probably a 404 status). If you run the `pactVerify` task with `-Ppact.showStacktrace` you will get the full stacktrace of the exception that the HTTP library raised.

I would guess that your provider does not accept POST requests to the root.

Currently, there is no way to disable this behaviour. I'll update the gradle plugin to not make the call if the state change URL is not provided, but this URL is normally needed to switch the provider to the correct state defined in the pact file.

**Comment by johnelf** [comment]
Hi Ron, so does that mean I need start another server to handle POST request to do some preparation work for state change? Because you know real server may not support that behavior.

**Comment by uglyog** [comment]
With the change I just made, you can leave it out if you don't need this behaviour.

However, if you do need to do some preparation work, then you will have to have a server that can accept the state via a POST and then perform the necessary work.

On my projects, I run the provider server (using DropWizard) with the test classpath which includes some extra URLs for the state change requests. But if you are unable to do that, you can create another server that has access to the necessary resources (like database) to handle the request.

Another option we could do (requires some work to implement), is to allow you to provide a gradle task to do the preparation work instead of the POST to a server.

**Comment by vishwt** [comment]
Is there any documentation or examples on how to use States??

**Comment by uglyog** [comment]
@vishwt information on provider states can be found here:

https://github.com/realestate-com-au/pact/wiki/Provider-states
https://github.com/DiUS/pact-jvm/tree/master/pact-jvm-provider-gradle#provider-states

**Comment by vishwt** [comment]
We are using pact jvm spring mvc
https://github.com/realestate-com-au/pact-jvm-provider-spring-mvc

How do we define stateChange URL here? State Change has to be always a URL?

**Comment by uglyog** [comment]
Currently, the only way to effect state changes is via a URL POST to the provider. For my projects (which I use Dropwizard), I have defined an additional administration URL that is only available when run with the test classpath. You could do something similar with spring-mvc.

**Comment by acooper267** [comment]
I'm using the maven plugin to verify pacts from a broker.
My provider has 4 different states. How do I set the statechangeurl and configure the plugin for all these states?

<stateChangeUrl>http://localhost:8080/tasks/pactStateChange</stateChangeUrl>


@State("A user with the role EXTERNAL_USER")
	public void stateExternalUser() {
		when(securityContext.isUserInRole("EXTERNAL_USER")).thenReturn(true);
	}
	
	@State("A user with the role INTERNAL_USER")
	public void stateInternalUser() {
		when(securityContext.isUserInRole("INTERNAL_USER")).thenReturn(true);
	}

**Comment by uglyog** [comment]
@acooper267 The state change URL will receive the state as a parameter. You can configure it as either a query parameter or have it as a JSON body.

**Comment by acooper267** [comment]
@uglyog I don't understand your answer.  Can you answer my question according to my example.

**Comment by uglyog** [comment]
The verifier will execute this request before the actual request:

```
POST  http://localhost:8080/tasks/pactStateChange
Content-Type: application/json
{
    "state": "A user with the role EXTERNAL_USER"
}
```

**Comment by acooper267** [comment]
Sorry if my question wasn't clear. I was asking what should I put in the pom xml file for the stateChangeUrl property. Should I replace "tasks/pactStateChange" with the text from my state? Do I need a stateChangeUrl property for each state? 

< stateChangeUrl > http://localhost:8080/tasks/pactStateChange </ stateChangeUrl >


based on my example pact test

@State("A user with the role EXTERNAL_USER")
 public void stateExternalUser() {
 when(securityContext.isUserInRole("EXTERNAL_USER")).thenReturn(true);
 }
@State("A user with the role INTERNAL_USER")
public void stateInternalUser() {
	when(securityContext.isUserInRole("INTERNAL_USER")).thenReturn(true);
}

**Comment by uglyog** [comment]
You just put the URL to be invoked as the `stateChangeUrl`. The actual state value from the pact file will be passed in as either query parameters or in the body of the POST request. Think of it as a callback mechanism that is invoked with the state from the pact file before the actual interaction is verified.

**Comment by varsha72** [comment]
Hi Team, 

 I am not able to publish the verification results to Pact Url,

could anyone look into the issue please, I am trying both using junit5 and using gradle

here is my class, removing d pact url and provider and consumer name for safety



import au.com.dius.pact.provider.PactVerifyProvider;
import au.com.dius.pact.provider.junit.Consumer;
import au.com.dius.pact.provider.junit.Provider;
import au.com.dius.pact.provider.junit.State;
import au.com.dius.pact.provider.junit.VerificationReports;
import au.com.dius.pact.provider.junit.loader.PactBroker;
import au.com.dius.pact.provider.junit.loader.PactUrl;
import au.com.dius.pact.provider.junit5.HttpTestTarget;
import au.com.dius.pact.provider.junit5.PactVerificationContext;
import au.com.dius.pact.provider.junit5.PactVerificationInvocationContextProvider;
import com.ihsmarkit.nextgen.analyticsservice.NextGenAnalyticsAPI;
import org.junit.ClassRule;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.TestTemplate;
import org.junit.jupiter.api.extension.ExtendWith;

@Provider("providername")
@Consumer("consumer name")
@PactUrl(urls = {"pacturl"})
@VerificationReports({"console", "markdown"})
public class ProviderTests {


    @BeforeEach
    void before(PactVerificationContext context) {
        context.setTarget(new HttpTestTarget("localhost", 8282, "/"));
        System.setProperty("pact.verifier.publishResults", "true");
        System.setProperty("pact.provider.version", "0.0.0");
    }


    @TestTemplate
    @ExtendWith(PactVerificationInvocationContextProvider.class)
    void pactVerificationTestTemplate(PactVerificationContext context) {
        context.verifyInteraction();
    }

    @State("order should be present")
    public void method1() throws InterruptedException {
    }

    @State("order related to deal id  should be present")
    public void method2() {
    }

}
 
pact{
    serviceProviders{
// You can define as many as you need, but each must have a unique name
        provider1 {
            // All the provider properties are optional, and have sensible defaults (shown below)
            protocol = 'http'
            host = 'localhost'
            port = 8282
            path = '/'

            test {
                systemProperties['pact.verifier.publishResults'] = "true"
            }

            // Again, you can define as many consumers for each provider as you need, but each must have a unique name
            hasPactWith('Analytics-BFF') {

                // currently supports a file path using file() or a URL using url()
                pactSource = url('pacturl')

            }
            reports {
                defaultReports() // adds the standard console output
                markdown // report in markdown format
                json // report in json format
            }

        }
    }
}

**Comment by uglyog** [comment]
@varsha72 the `test` block must be outside the `pact` block.
